### PR TITLE
fix/arc_nonsense

### PR DESCRIPTION
**Pull request by martindevans**
Previously the network system handled all packets as `Arc<Packet>`. This was unnecessary for many packets (cloning is cheap), slow (Arcs are expensive) and a pain to work with (e.g. matching through Arc). Removed that, packets can _internally_ use Arc where they contain expensive to clone things (e.g. `WorldTileData` now contains an `Arc<MeshData>`).

Packets were previously delivered inside a `ReceiveEvent`, which every system had to filter to a packet event, and then had to filter again to packets that system cared about. This was horrible to write (and also slow, lots of unnecessary filtering). This PR adds dispatching of packets out to a per-packet-type event, which systems can subscribe to.

### CODE REVIEW

**Review by ocornoc — APPROVED** [review]
It'd be nice to put in the docs of packet the caveats of `Arc`s in serde, but looks good.

**Comment by martindevans** [comment]
The caveats of Serde+Arc aren't particularly relevant to packet serialisation - since a packet will never be serialised and then deserialised on the same machine it's natural to assume a copy will be made (after all: what else could happen).

File: crates/bounded_planet_core/src/bin/server/main.rs [hunk 1]
```diff
@@ -123,9 +121,9 @@ pub struct PongLoggerState {
 fn log_pongs(mut state: ResMut<PongLoggerState>, receiver: ResMut<Events<ReceiveEvent>>) {
     for evt in state.event_reader.iter(&receiver) {
         if let ReceiveEvent::ReceivedPacket { data, .. } = evt {
-            if let Packet::Pong(Pong { timestamp }) = **data {
+            if let Packet::Pong(ref pong) = *data {
```
Comment (Jerald) [inline crates/bounded_planet_core/src/bin/server/main.rs]: I'm pretty sure this doesn't actually need to be fancy and use the ref and stuff...? Like with the others, it should be able to have the deref removed and it should all work out.

**Review by Jerald — CHANGES_REQUESTED** [review]

**Comment by Jerald** [comment]
My only thought on the larger direction of this is that now there is an additional concern when making new packets: the size and/or cost of cloning.

Previously you just made your packet be whatever it is you want them to be, and you didn't have to worry about accidentally slowing stuff down with large types and/or expensive clone implementations. Now, this is left up to the creator of the packet type, creating an opaque design footgun that's not obvious.

While in some ways I do see how this is "cleaner" and less stupid, I also think that it hid a lot of annoying design complexity that now needs to be addressed once again.

**Comment by martindevans** [comment]
Unfortunately I think that's just something we have to deal with. Throwing an `Arc` around for _every single packet_ really wasn't a very good idea!

Also of course to some extent you have to be thinking about the cost of cloning anyway. Since it's a packet that gets sent over the wire (which is a similar cost to cloning).

**Comment by ocornoc** [comment]
To be clear, you profiled this, right? This is an issue with any significant impact?

**Comment by martindevans** [comment]
No and probably not.

**Comment by Jerald** [comment]
No matter how we cut it, packets will need to live somewhere on the heap to outlast the individual systems they're from. Whether it's an `Arc` or some internal bevy magic is irrelevant, the cost is already paid. Because of this, I think it's not going to be such a concern that `Arc`s are going to actually limit is for real.

Basically, I think the advantages outweigh the _possible_ disadvantages. As some smart computer scientist said sometime: "premature optimization is the root of all evil"

File: crates/bounded_planet_core/src/bin/client/main.rs [hunk 2]
```diff
@@ -10,10 +10,7 @@ use bevy::{
     prelude::*,
     render::mesh::{Mesh, VertexAttribute}
 };
-use bounded_planet::{
-    camera::*,
-    networking::{events::*, packets::*, systems::*}
-};
+use bounded_planet::{camera::*, networking::{components::Connection, events::*, id::ConnectionId, packets::*, systems::*}};
```
Comment (Jerald) [inline crates/bounded_planet_core/src/bin/client/main.rs]: Unroll

File: crates/bounded_planet_core/src/land/systems.rs [hunk 3]
```diff
@@ -1,9 +1,6 @@
 use std::sync::Arc;
 use bevy::prelude::*;
-use crate::networking::{
-    events::{ReceiveEvent, SendEvent},
-    packets::{Packet, WorldTileDataRequest, WorldTileData}
-};
+use crate::networking::{id::ConnectionId, events::SendEvent, packets::{Packet, WorldTileData, WorldTileDataRequest}};
```
Comment (Jerald) [inline crates/bounded_planet_core/src/land/systems.rs]: Unroll

File: crates/bounded_planet_core/src/networking/client/plugin.rs [hunk 4]
```diff
@@ -1,30 +1,23 @@
...
 use url::Url;
 
-use crate::networking::{
-    crypto::SkipServerVerification,
-    events::{
-        ReceiveEvent,
-        SendEvent
-    },
-    systems::{
+use crate::networking::{crypto::SkipServerVerification, events::{NetworkError, SendEvent}, systems::{
```
Comment (Jerald) [inline crates/bounded_planet_core/src/networking/client/plugin.rs:20]: Unroll

File: crates/bounded_planet_core/src/networking/server/plugin.rs [hunk 5]
```diff
@@ -1,19 +1,14 @@
...
 use tokio::sync::mpsc::{unbounded_channel, UnboundedSender};
 use tracing::info;
 
-use crate::networking::{
-    events::{
-        ReceiveEvent,
-        SendEvent
-    },
-    systems::{
+use crate::networking::{events::{NetworkError, ReceiveEvent, SendEvent}, systems::{
```
Comment (Jerald) [inline crates/bounded_planet_core/src/networking/server/plugin.rs]: Unroll

**Review by Jerald — CHANGES_REQUESTED** [review]

**Pull request closed without merging**